File: 19-remove-nth-node-from-end-of-list/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class LinkListNode {
  val: number;
  next: LinkListNode | null;
  constructor(val?: number, next?: LinkListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function removeNthFromEnd(
  head: LinkListNode | null,
  n: number,
): LinkListNode | null {
  let cnt = 1;
  let now = head;
  while (!!now?.next) {
    cnt++;
    now = now.next;
  }

  const target = cnt - n - 1;
  if (target < 0) return head?.next || null;

  cnt = 0;
  now = head;
  while (cnt++ !== target) now = now?.next || null;

  (now as LinkListNode).next = now?.next?.next || null;
  return head;
}

File: 19-remove-nth-node-from-end-of-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* tmp = new ListNode(-1, head);
        ListNode* node = head;
        int sz = 0;
        while(node != nullptr) node = node->next, sz++;
        node = head;
        ListNode* prev = tmp;
        int i = 1;
        while(node != nullptr) {
            if(i == sz - n + 1) {
                prev->next = node->next;
                break;
            }
            prev = node, node = node->next, i++;
        }
        return tmp->next;
    }
};
